# DAMLAS MidTerm Exam  

11:00AM - 1:00PM(CT)
July 11, 2016   
Midterm


Data Analytics and Machine Learning at Scale

Target, Minneapolis

### Please insert your contact information here
Niki Deeny  
niki.deeny@target.com

In [1]:
import numpy as np
from __future__ import division

%reload_ext autoreload
%autoreload 2

# Exam Instructions

1. : Please insert Name and Email address in the first cell of this notebook
2. : Please acknowledge receipt of exam by sending a quick reply to exam email
3. : Review the submission form first to scope it out (it will take a 5-10 minutes to input your 
   answers and other information into this form): 

    + [Exam Submission form](http://goo.gl/forms/iN4XIXQyiH1iAdyD3) 

4. : Please keep all your work and responses in ONE (1) notebook only (and submit via the submission form)
5. : Please make sure that the NBViewer link for your Submission notebook works 
6. : Please submit your solutions and notebook via the following form:

     + [Exam Submission form](http://goo.gl/forms/iN4XIXQyiH1iAdyD3)

7. : For the midterm you will need access to MrJob and Jupyter on your local machines or on AWS to complete some of the questions (like fill in the code to do X).
8. : As for question types:
    + Knowledge test Programmatic/doodle (take photos; embed the photos in your notebook) 
    + All programmatic questions can be run locally on your laptop (using MrJob only) or on the cluster

9. : This is an open book exam meaning you can consult webpages and textbooks, class notes, slides etc. but you can not each other or any other person/group. Please complete this exam by yourself within the time limit. 

# Exam questions begins here

===Map-Reduce===

### MT1. Which of the following statememts about map-reduce are true? Check all that apply.   

(a) If you only have 1 computer with 1 computing core, then map-reduce is unlikely to help   
(b) If we run map-reduce using N computers, then we will always get at least an N-Fold speedup compared to using 1 computer   
(c) Because of network latency and other overhead associated with map-reduce, if we run map-reduce using N    computers, then we will get less than N-Fold speedup compared to using 1 computer   
(d) When using map-reduce with gradient descent, we usually use a single machine that accumulates the gradients from each of the map-reduce machines, in order to compute the paramter update for the iterion   

#####  MT1. a, b, c, d  # this is just a demo response
<span style="font-weight:bold; color:green;">__ANSWER__ This is Not necessary but it is always good way to cross check your logic. 
If you don't need the output to be sorted or aggregated, then there is no need for a reduce step. In other words, if all you're doing is applying some function to all data points and nothing else, then the reduce step is not needed.</SPAN>

#### MT1. Answer
A, B, D are true

===Order inversion===

### MT2. Suppose you wish to write a MapReduce job that creates  normalized word co-occurrence data form a large input text. 
To ensure that all (potentially many) reducers
receive appropriate normalization factors (denominators)
in the correct order in their input streams
(so as to minimize memory overhead), 
the mapper should emit according to which pattern:   

(a) emit (*,word) count   
(b) There is no need to use  order inversion here   
(c) emit (word,*) count   
(d) None of the above   

#### Answer MT2
C

===Map-Reduce===

### MT3. What is the input to the Reduce function in MRJob? Select the most correct choice.  

   
(a) An arbitrarily sized list of key/value pairs.    
(b) One key and a list of some values associated with that key  
(c) One key and a list of all values associated with that key.   
(d) None of the above   

#### Answer MT3
C

===Bayesian document classification===   
  
### MT4. When building a Bayesian document classifier, Laplace smoothing serves what purpose?   

(a) It allows you to use your training data as your validation data.   
(b) It prevents zero-products in the posterior distribution.  
(c) It accounts for words that were missed by regular expressions.    
(d) None of the above   

#### Answer MT4
B

===Bias-variance tradeoff===   

### MT5. By increasing the complexity of a model regressed on some samples of data,   it is likely that the ensemble will exhibit which of the following?   

(a) Increased variance and bias   
(b) Increased variance and decreased bias   
(c) Decreased variance and bias   
(d) Decreased variance and increased bias   

#### Answer MT5
B

Bias: the power of the model is too simple, Variance: power of model is too high, when you overfit the data then the predictions vary greatly when the backend data changes (traning data)

===Combiners===
  
### MT6. Combiners can be integral to the successful utilization of the Hadoop shuffle.  
This utility is as a result of (select the most correct answer only)  

(a) minimization of reducer workload 
(b) both (a) and (c)   
(c) minimization of network traffic    
(d) none of the above  

#### Answer MT6
B: Both A and C

## Pairwise similarity using K-L divergence

In probability theory and information theory, the Kullback–Leibler divergence 
(also information divergence, information gain, relative entropy, KLIC, or KL divergence) 
is a non-symmetric measure of the difference between two probability distributions P and Q. 
Specifically, the Kullback–Leibler divergence of Q from P, denoted DKL(P\‖Q), 
is a measure of the information lost when Q is used to approximate P:

For discrete probability distributions P and Q, 
the Kullback–Leibler divergence of Q from P is defined to be

    + KLDistance(P, Q) = Sum_over_item_i (P(i) log (P(i) / Q(i))      

In the extreme cases, the KL Divergence is 1 when P and Q are maximally different
and is 0 when the two distributions are exactly the same (follow the same distribution).

For more information on K-L Divergence see:

    + [K-L Divergence](https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence)

For the next three question we will use an MRjob class for calculating pairwise similarity 
using K-L Divergence as the similarity measure:

Job 1: create inverted index (assume just two objects)
Job 2: calculate/accumulate the similarity of each pair of objects using K-L Divergence

Download the following notebook and then fill in the code for the first reducer to calculate 
the K-L divergence of objects (letter documents) in line1 and line2, i.e., KLD(Line1||line2).

Here we ignore characters which are not alphabetical. And all alphabetical characters are lower-cased in the first mapper.

http://nbviewer.ipython.org/urls/dl.dropbox.com/s/9onx4c2dujtkgd7/Kullback%E2%80%93Leibler%20divergence-MIDS-Midterm.ipynb
https://www.dropbox.com/s/zr9xfhwakrxz9hc/Kullback%E2%80%93Leibler%20divergence-MIDS-Midterm.ipynb?dl=0

## Using the MRJob Class below  calculate the  KL divergence of the following two string objects.

In [20]:
%%writefile kltext.txt
1.Data Science is an interdisciplinary field about processes and systems to extract knowledge or insights from large volumes of data in various forms (data in various forms, data in various forms, data in various forms), either structured or unstructured,[1][2] which is a continuation of some of the data analysis fields such as statistics, data mining and predictive analytics, as well as Knowledge Discovery in Databases.
2.Machine learning is a subfield of computer science[1] that evolved from the study of pattern recognition and computational learning theory in artificial intelligence.[1] Machine learning explores the study and construction of algorithms that can learn from and make predictions on data.[2] Such algorithms operate by building a model from example inputs in order to make data-driven predictions or decisions,[3]:2 rather than following strictly static program instructions.

Overwriting kltext.txt


In [8]:
open("kltext.txt").read()

'1.Data Science is an interdisciplinary field about processes and systems to extract knowledge or insights from large volumes of data in various forms (data in various forms, data in various forms, data in various forms), either structured or unstructured,[1][2] which is a continuation of some of the data analysis fields such as statistics, data mining and predictive analytics, as well as Knowledge Discovery in Databases.\n2.Machine learning is a subfield of computer science[1] that evolved from the study of pattern recognition and computational learning theory in artificial intelligence.[1] Machine learning explores the study and construction of algorithms that can learn from and make predictions on data.[2] Such algorithms operate by building a model from example inputs in order to make data-driven predictions or decisions,[3]:2 rather than following strictly static program instructions.'

## MRjob class for calculating pairwise similarity using K-L Divergence as the similarity measure

Job 1: create inverted index (assume just two objects) <P>
Job 2: calculate the similarity of each pair of objects 

In [9]:
import numpy as np
np.log(3)

1.0986122886681098

In [10]:
!cat kltext.txt

1.Data Science is an interdisciplinary field about processes and systems to extract knowledge or insights from large volumes of data in various forms (data in various forms, data in various forms, data in various forms), either structured or unstructured,[1][2] which is a continuation of some of the data analysis fields such as statistics, data mining and predictive analytics, as well as Knowledge Discovery in Databases.
2.Machine learning is a subfield of computer science[1] that evolved from the study of pattern recognition and computational learning theory in artificial intelligence.[1] Machine learning explores the study and construction of algorithms that can learn from and make predictions on data.[2] Such algorithms operate by building a model from example inputs in order to make data-driven predictions or decisions,[3]:2 rather than following strictly static program instructions.

In [182]:
%%writefile kldivergence.py
from __future__ import division
from mrjob.job import MRJob
import re
import numpy as np

from math import log

class kldivergence(MRJob):
    
    # process each string character by character
    # the relative frequency of each character emitting Pr(character|str)
    # for input record 1.abcbe
    # emit "a"    [1, 0.2]
    # emit "b"    [1, 0.4] etc...
    def mapper1(self, _, line):
        index = int(line.split('.',1)[0])
        letter_list = re.sub(r"[^A-Za-z]+", '', line).lower()
        count = {}
        
        for l in letter_list:
            if count.has_key(l):
                count[l] += 1
            else:
                count[l] = 1
        for key in count:
            yield key, [index, count[key]*1.0/len(letter_list)]
            #KEY = the letter
            #INDEX = the which sentense its from
            #COUNT[key]*1.0 = the number of times it shows up in each
            #len(letter_list) number of letters (not unique) in each sentence
            #so it is outputting the letter, the sentence, and the likelihood of that letter showing up in that sentence
#             print "From mapper!", key, [index, count[key]*1.0/len(letter_list)]

    def reducer1(self, key, values):
        p = 0
        q = 0

        for v in values:
            if v[0] == 1:  #String 1
                p = v[1]
            else:          # String 2
                q = v[1]
    
        yield "key", (p*log(p/q)) # K-L Distance
        print "KL Divergence per letter", key, (p*log(p/q))

    #Aggegate components            
    def reducer2(self, key, values):
        kl_sum = 0
        for value in values:
            kl_sum += value
        yield "KLDivergence", kl_sum
            
    def steps(self):
        return [self.mr(mapper=self.mapper1,
                        reducer=self.reducer1),
                
                self.mr(reducer=self.reducer2)
               
               ]

if __name__ == '__main__':
    kldivergence.run()

Overwriting kldivergence.py


In [183]:
%reload_ext autoreload
%autoreload 2
from mrjob.job import MRJob
from kldivergence import kldivergence

#dont forget to save kltext.txt (see earlier cell)
mr_job = kldivergence(args=['kltext.txt'])
with mr_job.make_runner() as runner: 
    runner.run()
    # stream_output: get access of the output 
    for line in runner.stream_output():
        print mr_job.parse_output_line(line)

KL Divergence per letter a 0.0295721422713
KL Divergence per letter b -0.00163041522831
KL Divergence per letter c -0.00732786747342
KL Divergence per letter d 0.0164906236566
KL Divergence per letter e -0.0129926189574
KL Divergence per letter f 0.00674079918689
KL Divergence per letter g -0.00826965428728
KL Divergence per letter h -0.00992358514474
KL Divergence per letter i 0.00373655435066
KL Divergence per letter k 0.000733812807303
KL Divergence per letter l -0.0134916702888
KL Divergence per letter m -0.00829112158145
KL Divergence per letter n -0.021708593752
KL Divergence per letter o -0.00910212088756
KL Divergence per letter p -0.0094296551709
KL Divergence per letter r -0.0071047011805
KL Divergence per letter s 0.0907342592609
KL Divergence per letter t -0.0102420842309
KL Divergence per letter u 0.0147136183439
KL Divergence per letter v 0.0198601378947
KL Divergence per letter w 0.0176343237035
KL Divergence per letter x -0.00165393085746
KL Divergence per letter y 0.00

### MT7. Which number below is the closest to the result you get for KLD(Line1||line2)?   
(a) 0.7   
(b) 0.5   
(c) 0.2   
(d) 0.1  

#### Anser MT7
A

### MT8. Which of the following letters are missing from these character vectors?   
(a) p and t   
(b) k and q   
(c) j and q   

(d) j and f   

#### Answer MT8
C: J and Q

In [186]:
%%writefile kldivergence_smooth.py
from __future__ import division
from mrjob.job import MRJob
import re
import numpy as np

from math import log

class kldivergence_smooth(MRJob):
    
    # process each string character by character
    # the relative frequency of each character emitting Pr(character|str)
    # for input record 1.abcbe
    # emit "a"    [1, (1+1)/(5+24)]
    # emit "b"    [1, (2+1)/(5+24) etc...
    def mapper1(self, _, line):
        index = int(line.split('.',1)[0])
        letter_list = re.sub(r"[^A-Za-z]+", '', line).lower()
        count = {}
        
        # (ni+1)/(n+24)
        
        for l in letter_list:
            if count.has_key(l):
                count[l] += 1
            else:
                count[l] = 1
        for key in count:
            yield key, [index, (count[key]*1.0+1)/(len(letter_list)+24)] # Add the smoothing

    
    def reducer1(self, key, values):
        p = 0
        q = 0
        for v in values:
            if v[0] == 1:
                p = v[1]
            else:
                q = v[1]
                
        yield "key", (p*log(p/q)) # K-L Distance
        print "KL Divergence per letter", key, (p*log(p/q))

    # Aggregate components             
    def reducer2(self, key, values):
        kl_sum = 0
        for value in values:
            kl_sum = kl_sum + value
        yield "KLDivergence", kl_sum
            
    def steps(self):
        return [self.mr(mapper=self.mapper1,
                        reducer=self.reducer1),
                self.mr(reducer=self.reducer2)
               
               ]

if __name__ == '__main__':
    kldivergence_smooth.run()

Overwriting kldivergence_smooth.py


In [187]:
%reload_ext autoreload
%autoreload 2

from kldivergence_smooth import kldivergence_smooth
mr_job = kldivergence_smooth(args=['kltext.txt'])
with mr_job.make_runner() as runner: 
    runner.run()
    # stream_output: get access of the output 
    for line in runner.stream_output():
        print mr_job.parse_output_line(line)

KL Divergence per letter a 0.0271285770855
KL Divergence per letter b -0.00138634600173
KL Divergence per letter c -0.00693173000866
KL Divergence per letter d 0.0152918078694
KL Divergence per letter e -0.0124771140156
KL Divergence per letter f 0.00639606313847
KL Divergence per letter g -0.00797902003265
KL Divergence per letter h -0.00955497069346
KL Divergence per letter i 0.00310616930209
KL Divergence per letter k 0.000965278568718
KL Divergence per letter l -0.0128416309164
KL Divergence per letter m -0.00783050035972
KL Divergence per letter n -0.0206773746193
KL Divergence per letter o -0.00877209353032
KL Divergence per letter p -0.00973858492395
KL Divergence per letter r -0.00687026526917
KL Divergence per letter s 0.0835168653455
KL Divergence per letter t -0.00991976510841
KL Divergence per letter u 0.0137042758151
KL Divergence per letter v 0.0176835515088
KL Divergence per letter w 0.0140923225719
KL Divergence per letter x -0.00156610007194
KL Divergence per letter y 

### MT9. The KL divergence on multinomials is defined only when they have nonzero entries. 
For zero entries, we have to smooth distributions. Suppose we smooth in this way:    

(ni+1)/(n+24)   

where ni is the count for letter i and n is the total count of all letters.    
After smoothing, which number below is the closest to the result you get for KLD(Line1||line2)??   

(a) 0.08   
(b) 0.71     
(c) 0.02    
(d) 0.11   



#### Answer MT9
B

===Gradient descent===

### MT10. Which of the following are true statements with respect to gradient descent for machine learning, where alpha is the learning rate. Select all that apply   (stretch question)

 (a) To make gradient descent converge, we must slowly decrease alpha over time and use a combiner in the context of Hadoop.   
 (b) Gradient descent is guaranteed to find the global minimum for any function J() regardless of using a combiner or not in the context of Hadoop   
 (c) Gradient descent can converge even if alpha is kept fixed. (But alpha cannot be too large, or else it may fail to converge.) Combiners will help speed up the process.   
 (d) For the specific choice of cost function J() used in linear regression, there is no local optima (other than the global optimum).   

#### Answer MT10
B, C, D are true

## Weighted K-means

Below is MRJob code  for computing the KMeans of a 2-dimensional input data set. Modify this  MRJob to do the training at scale of a weighted K-means algorithm.

Weight each example as follows using the inverse vector length (Euclidean norm): 

    + weight(X)= 1/||X||, 

where ||X|| = SQRT(X.X)= SQRT(x1^2 + x2^2)

The weighted mean of a set of numbers of the form (x1 , x2), with corresponding weights (w1 , w2)  is computed from the following formula:

    + weightedMean = (w1 * x1 + w2 * x2) / (w1 + w2) 

Here X is vector made up of x1 and x2. For more information on the weighted mean, see https://en.wikipedia.org/wiki/Weighted_arithmetic_mean

Using the following data answer the following questions (download and save to this directory):

   + https://www.dropbox.com/s/ai1uc3q2ucverly/Kmeandata.csv?dl=0

### Check that the data is available in the local directory

In [199]:
!head Kmeandata.csv > head.csv

In [263]:
%%writefile MRKmeansIteration.py
from numpy import argmin, array, random
from math import sqrt   
from mrjob.job import MRJob
from mrjob.step import MRStep
import os

#Calculate find the nearest centroid for data point 
def MinDist(datapoint, centroid_points):
    datapoint = array(datapoint)
    centroid_points = array(centroid_points)
    diff = datapoint - centroid_points 
    diffsq = diff*diff
    # Get the nearest centroid for each instance
    minidx = argmin(list(diffsq.sum(axis = 1)))
    return minidx


class MRKmeansIteration(MRJob):
    k=3    
    def steps(self):
        return [MRStep(mapper_init = self.mapper_init, 
                       mapper=self.mapper,
                       combiner = self.combiner,
                       reducer=self.reducer)
               ]
#    def jobconf(self):
#        orig_jobconf = super(MRKmeansIteration, self).jobconf()        
#        custom_jobconf = {
#            'mapred.output.key.comparator.class': 'org.apache.hadoop.mapred.lib.KeyFieldBasedComparator',
#            'mapred.text.key.comparator.options': '-k1rn',
#            'mapred.reduce.tasks': '1',
#        }
#        combined_jobconf = orig_jobconf
#        combined_jobconf.update(custom_jobconf)
#        self.jobconf = combined_jobconf
#        return combined_jobconf

    def configure_options(self):
        """Add command-line options specific to this script."""
        super(MRKmeansIteration, self).configure_options()

        self.add_passthrough_option(
            '--centroidsFile', dest='centroidsFile', default="Centroids.txt", type='str',#type='int',
            help=('The location of the centroids file that contains the centroids '
                  ' that will be used for this iteration of KMeans Clustering. Default: %Centroids.txt'))

    #load centroids info from file
    def mapper_init(self):
        print "Current path:", os.path.dirname(os.path.realpath(__file__))
        self.centroid_points = [map(float,s.split('\n')[0].split(',')) for s in open(self.options.centroidsFile).readlines()]
        print "Centroids: ", self.centroid_points
        
    #load data and output the nearest centroid index and data point 
    def mapper(self, _, line):
        D = (map(float,line.split(',')))
        
        w = 1/(sqrt(D[0]**2 + D[1]**2)) # compute weight
        yield int(MinDist(D,self.centroid_points)), (D[0]*w,D[1]*w,w) # Change to weight
        
    #Combine sum of data points locally
    def combiner(self, idx, inputdata):
        sumx = sumy = num = 0
        for x,y,n in inputdata:
            num = num + n
            sumx = sumx + x
            sumy = sumy + y
        yield idx,(sumx,sumy,num)
        
    #Aggregate sum for each cluster and then calculate the new centroids
    def reducer(self, idx, inputdata): 
        sumx = sumy = num = 0
        for x,y,n in inputdata:
            num = num + n
            sumx = sumx + x
            sumy = sumy + y
        # send the centroids back to the driver
        yield idx,(sumx/num, sumy/num)

if __name__ == '__main__':
    MRKmeansIteration.run()

Overwriting MRKmeansIteration.py


In [136]:
%%writefile Centroids.txt
1.88191837648,0.761374538163
0.0194073517331,2.92156922133
0.334820407262,0.480496801461

Overwriting Centroids.txt


In [137]:
!head Centroids.txt   #fake cluster centroids

1.88191837648,0.761374538163
0.0194073517331,2.92156922133
0.334820407262,0.480496801461

In [264]:
#Test an iteration of KMeans
!python MRKmeansIteration.py --centroidsFile=Centroids.txt --file=Centroids.txt Kmeandata.csv

No configs found; falling back on auto-configuration
Creating temp directory /var/folders/r5/xnt6t6nd343fqlxyqjxqz3n50qpq8v/T/MRKmeansIteration.z084224.20160711.181758.485317
Running step 1 of 1...
Current path: /Users/z084224/Downloads
Centroids:  [[0.0935029243061, 0.0522456721659], [2.50120601597, 2.57660018845], [0.0407462704218, 0.0874330548455]]
Current path: /Users/z084224/Downloads
Centroids:  [[0.0935029243061, 0.0522456721659], [2.50120601597, 2.57660018845], [0.0407462704218, 0.0874330548455]]
Streaming final output from /var/folders/r5/xnt6t6nd343fqlxyqjxqz3n50qpq8v/T/MRKmeansIteration.z084224.20160711.181758.485317/output...
0	[2.759601420725678, -0.8601607205238444]
1	[2.972166472948007, 3.0350687877820035]
2	[-0.8842684642878555, 2.6515808763000153]
Removing temp directory /var/folders/r5/xnt6t6nd343fqlxyqjxqz3n50qpq8v/T/MRKmeansIteration.z084224.20160711.181758.485317...


In [265]:
%reload_ext autoreload
%autoreload 2
from itertools import chain
from numpy import random
from MRKmeansIteration import MRKmeansIteration


#Check whether the centroids have converged
def stop_criterion(centroid_points_old, centroid_points_new,T):
    oldvalue = list(chain(*centroid_points_old))
    newvalue = list(chain(*centroid_points_new))
    Diff = [abs(x-y) for x, y in zip(oldvalue, newvalue)]
    Flag = True
    for i in Diff:
        if(i>T):
            Flag = False
            break
    return Flag

#Generate initial centroids
centroid_points = []
k = 3
# the file that is broadcast to the workers before each iteration is started
centroidsFile = 'Centroids.txt'  
for i in range(k):
    centroid_points.append([random.uniform(-3,3),random.uniform(-3,3)])
with open(centroidsFile, 'w+') as f:
        f.writelines(','.join(str(j) for j in i) + '\n' for i in centroid_points)


# Create an instance of the KMeansIteration class
# This does all the work of assigning the data to centroids and recalculating the centroids 
# before shipping them back to the driver
#mr_job = MRKmeansIteration(args=['Kmeandata.csv', '--file=Centroids.txt',  '--k=3' ])
mr_job = MRKmeansIteration(args=['Kmeandata.csv', '--centroidsFile='+centroidsFile, '--file='+centroidsFile])

# Update centroids iteratively
i = 0
while(1):
    # save previous centoids to check convergency
    centroid_points_old = centroid_points[:]
    print "Starting KMeans Iteration "+str(i)+":"

    with mr_job.make_runner() as runner: 
        runner.run()

        # stream_output: get access of the output 
        for line in runner.stream_output():
            key,value =  mr_job.parse_output_line(line)
            centroid_points[key] = value

        # Update the centroids for the next iteration
        with open(centroidsFile, 'w') as f:
            f.writelines(','.join(str(j) for j in i) + '\n' for i in centroid_points)

        print "\n"
        i = i + 1
        if(stop_criterion(centroid_points_old,centroid_points,0.01)):
            break
print "Centroids\n"
print centroid_points

Starting KMeans Iteration 0:
Current path: /private/var/folders/r5/xnt6t6nd343fqlxyqjxqz3n50qpq8v/T/MRKmeansIteration.z084224.20160711.181803.445687/job_local_dir/0/mapper/0
Centroids:  [[-2.47308746955, -0.698071096448], [0.0433718238941, 1.48315312101], [-2.64057583265, 2.64423764392]]
Current path: /private/var/folders/r5/xnt6t6nd343fqlxyqjxqz3n50qpq8v/T/MRKmeansIteration.z084224.20160711.181803.445687/job_local_dir/0/mapper/1
Centroids:  [[-2.47308746955, -0.698071096448], [0.0433718238941, 1.48315312101], [-2.64057583265, 2.64423764392]]


Starting KMeans Iteration 1:
Current path: /private/var/folders/r5/xnt6t6nd343fqlxyqjxqz3n50qpq8v/T/MRKmeansIteration.z084224.20160711.181804.907908/job_local_dir/0/mapper/0
Centroids:  [[-0.109899942865, -1.17736476124], [3.11674276185, 2.4068664936], [-1.01491734809, 4.26519521422]]
Current path: /private/var/folders/r5/xnt6t6nd343fqlxyqjxqz3n50qpq8v/T/MRKmeansIteration.z084224.20160711.181804.907908/job_local_dir/0/mapper/1
Centroids:  [[-0.1

## KMEANS Questions:

### MT11. Which result below is the closest to the centroids you got after running your weighted K-means code for 10 iterations?   

(a) (-4.0,0.0), (4.0,0.0), (6.0,6.0)   
(b) (-4.5,0.0), (4.5,0.0), (0.0,4.5)  
(c) (-5.5,0.0), (0.0,0.0), (3.0,3.0)   
(d) (-4.5,0.0), (-4.0,0.0), (0.0,4.5)   

#### Answer MT11
A

Actual answers were (0,4), (6,6) (4,0)

### MT12. Using the result of the previous question, which number below is the closest  to the average weighted distance between each example and its assigned (closest) centroid?
The average weighted distance is defined as 
* sum over each data point i  (weight_i * MinDist(datapoint, centroid_points) )     /  (sum over i (weight_i))

where *weight_i* is defined as above:

* weight_i = weight(Xi)= 1/||Xi||, 

where ||Xi|| = SQRT(Xi.Xi)= SQRT(x1_i^2 + x2_i^2)


(a) 2.5   
(b) 1.5   
(c) 0.5   
(d) 4.0

#### Answer MT12
B

===Map-Reduce===

### MT13. Which of the following statements are true? Select all that apply.    
 (a)	  Since K-Means is an unsupervised learning algorithm, it cannot overfit the data, and thus it is always better to have as large a number of clusters as is computationally feasible.    
 (this is jus one way. you can also use random initialization) 
 (b)	  The standard way of initializing K-means is setting μ1=⋯=μk to be equal to a vector of zeros.    
 (c)	  For some datasets, the "right" or "correct" value of K (the number of clusters) can be ambiguous, and hard even for a human expert looking carefully at the data to decide.    
 (d)	  A good way to initialize K-means is to select uniformly at random K (distinct) examples from the training set and set the cluster centroids equal to these selected examples.    

#### Answer MT13
C, D are true

### MT14. Is there a map input format (for Hadoop or MRJob)?   

(a)  Yes, but only in Hadoop 0.22+.   
(b)  Yes, in Hadoop there is a default expectation that each record is delimited by an end of line charcacter and that key is the first token delimited by a tab character and that the value-part  is everything after the tab character.   
(c)  No,  when  MRJob INPUT_PROTOCOL = RawValueProtocol. In this case input is processed in format agnostic way thereby avoiding any type of parsing errors. The value is treated as a str, the key is read in as None.   
(d) Both b and c are correct answers.  

#### Answer MT14
D

### MT15. What happens if mapper output does not match reducer input?   
 
(a)  Hadoop API will convert the data to the type that is needed by the reducer.    
(b)  Data input/output inconsistency cannot occur. A preliminary validation check is executed prior to the full execution of the job to ensure there is consistency.    
(c)  The java compiler will report an error during compilation but the job will complete with exceptions.    
(d)  A real-time exception will be thrown and map-reduce job will fail.

#### Answer MT15
D

### MT16. Why would a developer create a map-reduce without the reduce step?   
 
(a)  Developers should design Map-Reduce jobs without reducers only if no reduce slots are available on the cluster.    
(b)  Developers should never design Map-Reduce jobs without reducers. An error will occur upon compile.    
(c)  There is a CPU intensive step that occurs between the map and reduce steps. Disabling the reduce step speeds up data processing.  
(d)  It is not possible to create a map-reduce job without at least one reduce step. A developer may decide to limit to one reducer for debugging purposes.   



#### Answer MT16
C

# END of Exam